### ISSUE TITLE
Tx: EIP-1559 V value of 26 should throw error

### ISSUE THREAD

**Issue by Hugo-Vincent** [ethereumjs/ethereumjs-monorepo#1646]
I'm involved in a project where I get an ETH DER-signature from a black box, so I cannot use the ethereumjs/tx sign function. I have to set the R, S and V values manually, and I use _processSignature for this. Oddly enough I found that besides 27 and 28, it also accepts 26.

``` typescript
import { FeeMarketEIP1559Transaction as EIP1559Tx } from '@ethereumjs/tx';
import Web3 from 'web3';
import BN from 'bn.js';

/**
 * Responsible for initializing and calling @ethereumjs/tx class instances, which allow to save and load, build and
 * sign transactions.
 * {@class EthLib}
 */
export class EthLib {

  /**
   * Test sign function to check recovered addresses.
   * @param {string} txHash
   * @param {string} privateKey
   * @param {string} publicKey
   * @param {string} address
   */
  sign(): void {
    
    // Setup
    const address = '0xbbcbba8adae660238c1398c52a23179d555f50c1';
    const publicKey = '0399f70cde67679823ffc7aa5a121947c7b10be6f985b93d8aaaec80d9944daad1';
    const privateKey = '6206e509e4ea505633081594b235a1c92b05135c0f886bfde789df968efe463a';
    const serializedTx = '02e70305800a82520894e44ef3a1713c2b140cd9a157d4a162db459b2daf8505d21dba0080c0808080';

    let tx = EIP1559Tx.fromSerializedTx(Buffer.from(serializedTx, 'hex'));
    tx = tx.sign(Buffer.from(privateKey, 'hex'));
    console.log(tx);
    let recoveredAddress = new Web3().eth.accounts.recoverTransaction(`0x${tx.serialize().toString('hex')}`);
    console.log(recoveredAddress + '\n' + address);

    // Change the V value (if tx.v was 1, then input 27 here, else 28)
    tx = tx._processSignature(
      tx.v.eq(new BN(1)) ? 27 : 28,
      Buffer.from(tx.r.toString('hex'), 'hex'),
      Buffer.from(tx.s.toString('hex'), 'hex'),
    );

    // Now we get a different address
    console.log(tx);
    recoveredAddress = new Web3().eth.accounts.recoverTransaction(`0x${tx.serialize().toString('hex')}`);
    console.log(recoveredAddress + '\n' + address);
    
    // Only 27 and 28 should be valid input values, 29 throws an error as expected.
    try {
      tx = tx._processSignature(29, Buffer.from(tx.r.toString('hex'), 'hex'), Buffer.from(tx.s.toString('hex'), 'hex'));
    } catch (e) {
      console.log(e);
    }
    
    // However, 26 doesn't throw an error. What's up?
    try {
      tx = tx._processSignature(26, Buffer.from(tx.r.toString('hex'), 'hex'), Buffer.from(tx.s.toString('hex'), 'hex'));
      console.log(' no error??');
      recoveredAddress = new Web3().eth.accounts.recoverTransaction(`0x${tx.serialize().toString('hex')}`);
      console.log(recoveredAddress + '\n' + address); // Same as when v = 28
    } catch (e) {
      console.log(e);
    }
  }
}

```

I assume this isn't supposed to be the case?
eip1559Transaction.ts @line 416:
```
        v: new BN(v - 27), // This looks extremely hacky: ethereumjs-util actually adds 27 to the value, the recovery bit is either 0 or 1.
```

So v should become -1, which is an invalid value, however the V value for this signing operation becomes 1 in the end.

**Comment by holgerd77** [comment]
Thanks a lot for reporting 😄, these kind of bugs are indeed sensitive and should be fixed!

Incidentally we just had adressed exactly the same issue in one of the last PRs - see #1606.

I wasn't 100% sure if this would apply to this use case but have tested and can confirm that this now also throws:

```shell
tx = tx._processSignature(26, Buffer.from(tx.r!.toString('hex'), 'hex'), Buffer.from(tx.s!.toString('hex'), 'hex'));
/ethereumjs-monorepo/packages/util/dist/bytes.js:157
            throw new Error(`Cannot convert negative BN to buffer. Given: ${v}`);
            ^

Uncaught Error: Cannot convert negative BN to buffer. Given: -1
    at toBuffer (/ethereumjs-monorepo/packages/util/src/bytes.ts:176:13)
    at new BaseTransaction (/ethereumjs-monorepo/packages/tx/src/baseTransaction.ts:90:24)
    at new FeeMarketEIP1559Transaction (/ethereumjs-monorepo/packages/tx/src/eip1559Transaction.ts:186:5)
    at Function.fromTxData (/ethereumjs-monorepo/packages/tx/src/eip1559Transaction.ts:89:12)
    at FeeMarketEIP1559Transaction._processSignature (/ethereumjs-monorepo/packages/tx/src/eip1559Transaction.ts:411:40)
    at /ethereumjs-monorepo/packages/vm/<repl>.ts:1:9
    at Script.runInThisContext (node:vm:129:12)
```

Changes are currently on `master` and will be released along the next release round which is already in preparation - see #1640 - planned to be released within the next 1-3 weeks. 🙂

**Comment by holgerd77** [comment]
Leaving this open here a bit, in cases where people need to use private functionality I always wonder a bit if something is wrong with our API, especially in this case because I think we might have had at least one similar case before here.

@ryanio (everyone), would it make sense to publicly expose this functionality? 🤔

Would there be a more descriptive name for that, I find the `processSignature` thing not so self-explanatory, especially on the "I get a new tx back side"?

First suggestion:

`tx.getSignedTx(v, r, s)`

I guess `tx.fromSignature(v, r, s)` might clash a bit too much with our static constructor naming?

@Hugo-Vincent what are your thoughts about this?

**Comment by ryanio** [comment]
hmm, just thinking off the top of my head, how about `tx.signManual(v, r, s)`, to complement the existing `tx.sign(privKey)`

**Comment by holgerd77** [comment]
Hmm, I also thought about overloading `tx.sign()`, but both of this is actually misleading, since the signature has already been created, so this is technically wrong. How about `tx.applySignature()`?

**Comment by ryanio** [comment]
> but both of this is actually misleading, since the signature has already been created, so this is technically wrong.

I'm not sure I understand what you mean 🤔  I do think we should return a new tx with the applied values and not modify the tx in-place, to maintain immutability.

**Comment by holgerd77** [comment]
Ah no, I didn't want to put the structure in question with returning a new tx, this is just about proper naming. 

So my reasoning was: "signing" happens when this `ecsign()` function is called, then this process is over. So the argument was that "signing" is actually "over" when you already have the `v`, `r`, `s` values and to call this something with `sign*` would actually be misleading.

Not sure if this holds though on second thought.

Maybe your name is good as well and can be grasped relatively easily.

Then we should also make another round and also take the option to use the `tx.sign()` method directly into account, so something like:

```typescript

interface signature = { // or maybe directly `ecdsaSignature` for future compatibility
  v: BN,
  r: BN,
  s: BN
}

sign(keyOrSignature: Buffer | Signature)
```

Hmm, but also not sure now seeing it. Maybe the other option with a separate name is better. Some aspect: I guess we need to copy over most of the code from `sign`, on first sight I think we even need to take this hack-code.

WDY(G)T?

**Comment by Hugo-Vincent** [comment]
@holgerd77 Hello, great I see you had already discovered and addressed the bug :)

About manually setting the values, to my knowledge this only happens when people do not use @ethereumjs/tx sign() function, because they do not have access to the private key. So in this case, we receive the (R, S) value from the signing party while we keep the unsigned transaction stored away in a serialized string.

If I want to "inject" the signature into the unsigned transaction, I take the following steps:
- `const tx = FeeMarketEIP1559Transaction.fromSerializedTx(Buffer.from(savedTx, 'hex'))`
- `const signed tx = tx._processSignature(27, R, S)`
- Then I use `Web3.eth.accounts.recoverTransaction()` to recover the sending address from the signature. 
- If V = 27 did not produce the original sending address, I  set V to 28 and do `const signed tx = tx._processSignature(28, R, S)`

`_processSignature()` is actually just `FeeMarketEIP1559Transaction.fromTxData(txData )` again. So in worst case I'm building the transaction 3 times. Preferable I would like to build it only 1 time, and just set R, S and V manually as I please.
I understand why they have become `readonly` though. It's a safety measure as there's no need at all to change those values if you just use the @ethereumjs/tx `sign()` function.

So I was planning one of three solutions:
1. Extend your class with another class where I would re-declare R, S and V, but simply remove the `readonly` property which would overwrite the superclass.  Then I wouldn't need `_processSignature()` either way.

2. Create a static constructor `FeeMarketEIP1559Transaction.fromSignedTxData(txData, R, S)`, which would create the transaction 1 time with R, S included. Redeclare V without `readonly` and set it after correct pubkey was found.

3. Create a static constructor `FeeMarketEIP1559Transaction.fromSignedTxData(txData, R, S, V)`, which would create the transaction 1 time with R, S, V included. (Determine V beforehand using a DER-parser)

In my own code I have called it `injectSignature(R, S)` which calls @ethereumjs/tx, because I'm just injecting values, so I thought that was an appropriate name.

**Comment by ryanio** [comment]
@Hugo-Vincent thanks for the detailed write up! Is there a reason why you don't pass v,r,s into fromTxData along with the other tx values? That would achieve the same result as your proposed "fromSignedTx"

The reason all the properties in the tx class are readonly is to maintain immutability and unwanted side effects from modifying txs in-place, so we likely want to keep that design ideology in mind.

**Comment by Hugo-Vincent** [comment]
@ryanio Well by the time I receive R S V, I only have the serialized Tx string stored in a DB, so I need to call `FeeMarketEIP1559Transaction.fromSerializedTx()` to decode it into an object first. Then I can use that to extract the values and, together with R S and V, create a `FeeMarketEIP1559TxData` object, which I can then use to call `FeeMarketEIP1559Transaction.fromTxData()` yes, as you say.

But then I'm better off using `_processSignature()` after deserializing, because it does exactly that: extract the values from the tx, and put it in a  `FeeMarketEIP1559TxData` object together with R, S and V, and then put that in the `fromTxData()` call. Which is what I currently do, which I call "injectSignature()" in my api that calls @ethereumjs/tx `_processSignature()`.

It's not really a problem, it's just a bit of a roundabout way of doing things. Some would probably call the 3 solutions I mentioned to be insignificant micro-optimizations. It's just that in my case, I have to take scalability into account, as we might have to process thousands or tens of thousands of signatures an hour, which is where small optimizations may make a more significant impact.

So for me personally it would be worth it to extend the FeeMarketEIP1559Transaction class and make a simple function that modifies the `FeeMarketEIP1559Transaction.fromSerializedTx()` that also takes R S and V and creates everything in 1 go. And I agree strongly with the immutability ideology.
Anyway, just my thoughts, whether this is applicable to a wider audience of @ethereumjs/tx users I don't know :)

**Comment by ryanio** [comment]
thanks for sharing, it's always helpful to get perspective on how our libraries are used :)

so yeah `_processSignature(v,r,s)` basically has the API we had in mind for what `signManual(v,r,s)` would be then. It would just be more easily usable and visible if we make it part of the public tx API.

I think `fromSerializedTx(serializedUnsignedTx)` then `_processSignature(v,r,s)` for your specific use case works well.

**Comment by Hugo-Vincent** [comment]
@ryanio No problem, and yes I agree.

**Comment by holgerd77** [comment]
Addressed in #3238, will close.
